designs/0005-static-matrices.md

```diff
@@ -0,0 +1,444 @@
+that subsets of `varmat` variables cannot be assigned efficiently. The efficiency loss
+was so great in the original design (requiring a copy of the entire `varmat`) that this
+idea was discarded completely. There is a strategy available now to assign to subsets
+of `varmat` variables that is more efficient than this.
+
+The idea is that when part of a `varmat` is assigned to, the values in that part of the
+`varmat` are saved into the memory arena before being overwritten, and a call is pushed
+onto the chaining stack that will restore the values back from the arena and set the
+respective adjoints of the `varmat` entries to zero.
+
+The assumptions that makes this work is after a subset of a `varmat` is overwritten,
+that part of the `varmat` cannot be used in any further calculations, and so the adjoints
+of that part of the matrix will be zero in the reverse pass at the time of assignment.
+
+This means that assignment into a `varmat` variable would make it so that views
+point into that part of the `varmat` variable's memory no longer point to the expected
+variables.
+
+# Stanc3 implementation
+[stanc3-implementation]: #stanc3-implementation
+
+The stanc3 implementation goal is to implement `varmat` types in a way that
+is invisible to the user. This means the compiler decides how Stan vectors, row
+vectors, and matrices (or arrays of these types) are implemented automatically,
+either picking an appropriate `varmat` or a `matvar` type, and these decisions should
+be made in a way to avoid making programs slower than they would have been
+with just `matvar` types. For brevity, any Stan vector, row vector, or matrix
+type will be referred to from hereon as just a matrix type. Consider a linear
+regression model:
+
+```stan
+data {
+  int N;
+  int K;
+  matrix[N, K] X;
+  vector[N] y;
+}
+
+parameters {
+  vector[K] b;
+  real<lower = 0.0> sigma;
+}
+
+model {
+  vector[N] mu = X * b;
+  target += normal_lpdf(y | mu, sigma);
+}
+```
+
+In this program there are four named matrix variables, `X`, `y`, `b`, and `mu`.
+
+It is immediately evident that because there are no autodiff types in the
+`data` block, all the matrix types in there are implemented with doubles in
+Eigen, so there is no `matvar` or `varmat` to worry about. The same thing
+applies for `transformed data` and the `generated quantities` block.
+
+This leaves `b` and `mu`, matrix variables defined in the `parameters` and
+`model` block. Any matrix variable defined in a `transformed parameters` block
+would be decided the same way.
+
+The algorithm for figuring out the underlying matrix types for these variables
+is as follows:
+
+1. Define all user-defined functions only for `matvar` types, and all
+internally defined variables and returned variables are `matvar` types as well.
+
+2. Assume every named and unnamed matrix variable in the `parameters`,
+`transformed parameters` and `model` blocks use `varmat` types
```

Comment:
Yes, currently all Stan Math functions that do not take in an Eigen-type but do return it, return a matvar Eigen-type. I would say once we do support returning varmat in those, they would be separated via a template. Other options like suffixes would cause a ton of issues in the compiler.

Reply:
> Not sure what these "unnamed matrix variables" are?

Temporaries, like:

```
matrix[5, 5] A = B * C + D;
```

Then there's probably a temporary variable for `B * C`, and it has no name.

> It is certainly possible for stanc3 to figure out the optimal return type for them.

For functions defined in Math, the list of function names + arguments -> return types will be given. Stanc3 won't have to do anything here.

For user-defined functions, in the current proposal, all return types are matvar.

Once we start deducing types for user-defined functions, then we could hopefully decide the return type using similar rules as we have everywhere else (use varmat unless there is a rule that says not to).

I hope we can avoid having the return type be part of the type of the function. It seems nice that we can take name + argument types and compute a return type. Functions with multiple return types sound annoying.